Add Tests for ConfirmImpact Ratings

vexx32 opened the pull request:
## PR Summary

/cc @iSazonov 

Adds aspect to tests in `test/powershell/engine/Basic/DefaultCommands.Tests.ps1` that checks default cmdlets' reported `ConfirmImpact` level.

Also fixes a related logical error in `/tests/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1` where the test attempts to query the wrong property on the class object, and then tests for `ConfirmImpact` rating without implementing `ShouldProcess`.

Related: #8209 
These tests will be modified in #8209 when needed.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

vexx32 committed:
Fix broken test to properly apply attribute properties it was expecting

vexx32 committed:
Add file to test ConfirmImpact ratings of commands
Also test that number of cmdlets matches what is available.

vexx32 committed:
Use more direct method of determining the commands available
Might be needed since unix doesn't have some commands

soft-fail for missing cmdlets on unix OSes for ConfirmImpact checks

vexx32 committed:
Add more robust test for command availability

Fix typos

Remove unnecessary sort

vexx32 committed:
Attempt to fix weird issues

Use proper import of default modules

Remove unneeded check

Fix spacing

vexx32 commented on test/powershell/engine/Basic/DefaultConfirmImpact.Tests.ps1:
```diff
@@ -0,0 +1,268 @@
+
+        # ConfirmImpact.Low
+            @{ Cmdlet = 'Disable-PSSessionConfiguration'; ConfirmImpact = 'Low' }
+            @{ Cmdlet = 'Unregister-PSSessionConfiguration'; ConfirmImpact = 'Low' }
```
Roger that! 😄

iSazonov commented on test/powershell/engine/Basic/DefaultConfirmImpact.Tests.ps1:
```diff
@@ -0,0 +1,271 @@
+    It '<Cmdlet> should have a ConfirmImpact rating of <ConfirmImpact>' -TestCases $DefaultCommands {
+        param($Cmdlet, $ConfirmImpact)
+
+        $Cmdlet | Should -BeIn $CommandList
```
I think we can remove the check - it is not related the PR.

vexx32 replied:
I was hoping to somehow verify that all default commands are being properly tested in this fashion... but it doesn't seem to like what I'm trying to do at the moment. 😄

iSazonov commented on test/powershell/engine/Basic/DefaultConfirmImpact.Tests.ps1:
```diff
@@ -0,0 +1,256 @@
+            @{ Cmdlet = 'Write-Output'; ConfirmImpact = 'Medium' }
+            @{ Cmdlet = 'Write-Progress'; ConfirmImpact = 'Medium' }
+            @{ Cmdlet = 'Write-Verbose'; ConfirmImpact = 'Medium' }
+            @{ Cmdlet = 'Write-Warning'; ConfirmImpact = 'Medium' }
```
I'd reformat the array to align like:
```suggestion
            @{ Cmdlet = 'Write-Progress';          ConfirmImpact = 'Medium' } 
            @{ Cmdlet = 'Write-Verbose';           ConfirmImpact = 'Medium' }
            @{ Cmdlet = 'Write-Warning';          ConfirmImpact = 'Medium' }
```

vexx32 replied:
I had it like this at one point, but I guess I must have had to copy paste it at some point an VS Code decided it wasn't needed. Fixed! 😄

iSazonov commented:
@SteveL-MSFT @adityapatwardhan Could you please review these new tests?

SteveL-MSFT commented on test/powershell/engine/Basic/DefaultConfirmImpact.Tests.ps1:
```diff
@@ -0,0 +1,256 @@
+Describe 'Default Cmdlet ConfirmImpact Ratings' -Tags 'CI' {
```
Please copy the copyright notice to the top of this file

SteveL-MSFT commented on test/powershell/engine/Basic/DefaultConfirmImpact.Tests.ps1:
```diff
@@ -0,0 +1,256 @@
+            "Microsoft.PowerShell.Security"
+            "Microsoft.PowerShell.Host"
+            "Microsoft.PowerShell.Diagnostics"
+            "PSWorkflow"
```
Why is PSWorkflow here?

iSazonov replied:
The test should works on Windows PowerShell and contains not ported Windows PowerShell cmdlets - so we can see that is difference PowerShell Core and Windows PowerShell..

SteveL-MSFT commented on test/powershell/engine/Basic/DefaultConfirmImpact.Tests.ps1:
```diff
@@ -0,0 +1,256 @@
+            "Microsoft.PowerShell.Host"
+            "Microsoft.PowerShell.Diagnostics"
+            "PSWorkflow"
+            "Microsoft.WSMan.Management"
```
None of the cmdlets from this module are in the list below

SteveL-MSFT commented on test/powershell/engine/Basic/DefaultConfirmImpact.Tests.ps1:
```diff
@@ -0,0 +1,256 @@
+            "Microsoft.PowerShell.Management"
+            "Microsoft.PowerShell.Security"
+            "Microsoft.PowerShell.Host"
+            "Microsoft.PowerShell.Diagnostics"
```
Cmdlets for this module are not included below

SteveL-MSFT commented on test/powershell/engine/Basic/DefaultConfirmImpact.Tests.ps1:
```diff
@@ -0,0 +1,256 @@
+            "Microsoft.WSMan.Management"
+            "Microsoft.PowerShell.Core"
+        )
+        Import-Module -Name $moduleList -ErrorAction SilentlyContinue
```
Actually, you shouldn't have to explicitly have this list of modules and import them as auto-module loading will take care of that for you

vexx32 replied:
It wasn't. 🤷‍♂️ 

Not sure why, though.

iSazonov replied:
This did not work in DefaultsCommands.Tests.ps1 too.

SteveL-MSFT commented on test/powershell/engine/Basic/DefaultConfirmImpact.Tests.ps1:
```diff
@@ -0,0 +1,256 @@
+    It '<Cmdlet> should have a ConfirmImpact rating of <ConfirmImpact>' -TestCases $DefaultCommands {
+        param($Cmdlet, $ConfirmImpact)
+
+        $Type = (Get-Command -Name $Cmdlet).ImplementingType
```
Thinking about this, it would be better to just add a new column to the table in https://github.com/PowerShell/PowerShell/blob/master/test/powershell/engine/Basic/DefaultCommands.Tests.ps1 and validate the confirmimpact in that test file

vexx32 replied:
If I'm going to have to add to that file, I'd like to overhaul it. The CSV type format is ridiculously hard to read, probably review, add to, and I'm sure it's not fun to maintain.

Most of these sorts of tests that I've seen thus far opt for an array of hashtables rather than a big block of CSV text. I feel like that might also be a better option here?

iSazonov replied:
We could merge these tests. 
I think we don't improve readability if we convert to hashtable - we get very long line or we'll be forced to collapse to multiline.
So I'd simply add the new column.

vexx32 replied:
Sounds good... The best way for me to do this would be programmatically from a current version of PS, but that'll butcher the spacing... I suppose I can just add that back in... 🤷‍♂️ yeah it'll work.

iSazonov replied:
VS Code support to edit _columns_.

vexx32 committed:
Update defaultCommands tests to include ConfirmImpact

Remove unneeded additional test file

Update tests with -Because for easier debugging

Remove tabs for test CSV string

Reorganise test

I think I fixed it
(famous last words, eh?)

Add back missing Present flags for PSHostProcess items

vexx32 commented:
@iSazonov built ConfirmImpact tests into the existing file per @SteveL-MSFT's recommendations. Haven't tested it just yet, but it should be sound. I'll let CI check it while I get some sleep and figure out any issues tomorrow.

Also did a rebase due to some outstanding merge conflicts & resolved those issues.

@SteveL-MSFT I notice that there are references to `$FullCLR` in this file; are these vestiges from Windows PowerShell that can be removed unilaterally?

iSazonov commented:
$FullCLR helps to see the differences from Windows PowerShell, helps to document. Also we can run the test with Windows PowerShell to keep it up-to-date.

vexx32 committed:
Let test show cmdlet name for failing confirmimpact items

vexx32 commented:
Okay, it took me a bit to get that working as it perhaps should, but there we go. 😄

iSazonov commented:
@vexx32 Please resolve the merge conflict.

@SteveL-MSFT Please update your code review.

vexx32 commented:
There we go! 😄

adityapatwardhan commented:
@vexx32 Please have a look at the test failures on macOS and Linux CI.

Also, please update the PR description to remove reference to : `DefaultConfirmImpact.Tests.ps1`

vexx32 committed:
Add $CoreUnix specifier to New-PSSessionOption

vexx32 commented:
@adityapatwardhan Tests and PR description updated!

iSazonov commented:
@vexx32 Please resolve the merge conflit.

@SteveL-MSFT Please update your code review.

iSazonov commented:
@vexx32 Please resolve the merge conflit.

@SteveL-MSFT Please update your code review.

iSazonov commented:
@vexx32 Please resolve the merge conflit.

@SteveL-MSFT Please update your code review.

iSazonov commented:
@vexx32 Please resolve the merge conflit.

@SteveL-MSFT Please update your code review.

iSazonov commented:
@vexx32 Please resolve the merge conflit.

@SteveL-MSFT Please update your code review.

iSazonov commented:
@vexx32 Please resolve the merge conflit.

@SteveL-MSFT Please update your code review.

vexx32 commented:
I think that takes care of it. Managed to figure out which commands have been added since last commit, so hopefully that should be sufficient, but I'm gonna  wait and see if CI reckons I missed something and fix it as it happens. 😄

iSazonov commented:
@vexx32 Please resolve the merge conflit.

@SteveL-MSFT Please update your code review.

vexx32 commented:
Merge conflict is sorted /cc @SteveL-MSFT  @iSazonov :smile:

SteveL-MSFT reviewed (approved):
Sorry for the delay, been busy and then got backlogged on 90 GitHub notifications but finally got to this one

vexx32 commented:
All good, I imagine it gets fun around the holidays over there. 😄
